src/map/utils/puppetutils.cpp

```diff
@@ -778,4 +778,50 @@ namespace puppetutils
         }
     }
 
+    void PreLevelRestriction(CCharEntity* PChar)
+    {
+        CAutomatonEntity* PAutomaton = PChar->PAutomaton;
+        if (PAutomaton)
+        {
+            for (int i = 0; i < 12; i++)
+            {
+                uint8 attachment = PAutomaton->getAttachment(i);
+
+                if (attachment != 0)
+                {
+                    CItemPuppet* PAttachment = (CItemPuppet*)itemutils::GetItemPointer(0x2100 + attachment);
+
+                    // Attachment scripts may have custom unequip logic that needs to run before the restriction is applied
+                    // If they were to delMod after the restriction is applied, under/overflow may occur.
+                    // This will also clear the localVars holding previously applied modifiers
+                    luautils::OnAttachmentUnequip(PAutomaton, PAttachment);
+                }
+            }
+        }
+    }
+
+    void PostLevelRestriction(CCharEntity* PChar)
+    {
+        CAutomatonEntity* PAutomaton = PChar->PAutomaton;
+
+        if (PAutomaton)
+        {
+            for (int i = 0; i < 12; i++)
+            {
+                uint8 attachment = PAutomaton->getAttachment(i);
+
+                if (attachment != 0)
+                {
+                    CItemPuppet* PAttachment = (CItemPuppet*)itemutils::GetItemPointer(0x2100 + attachment);
```

Comment:
```suggestion
                    CItemPuppet* PAttachment = dynamic_cast<CItemPuppet*>(itemutils::GetItemPointer(0x2100 + attachment));
                    if (PAttachment)
                    ...
```

Reply:
If it must be a hard cast, use `static_cast<CItemPuppet*>`, we prefer to use dynamic_casting and checking the result however